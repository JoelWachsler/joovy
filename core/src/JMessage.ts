import { Message, MessageOptions, MessagePayload } from 'discord.js'
import { defer, filter, from as rxFrom, map, mapTo, mergeMap, Observable, of } from 'rxjs'
import { QueueReactions } from './commands/impl/Queue'

/**
 * Facade around Message for easier test implementations.
 */
export default interface JMessage {
  channelId: string
  author: {
    id: string
    username: string
    bot: boolean
  }
  content: string
  clearReactions: Observable<JMessage>
  reactions: Observable<JReaction>
  messageKey: MessageKey
  edit(update: MessageContent): Observable<JMessage>
  react(reaction: string): Observable<JMessage>
  send(message: MessageContent): Observable<JMessage>
}

export interface MessageKey {
  channelId: string
  messageId: string
}

export type MessageContent = string | MessagePayload | MessageOptions

export type JReaction = string

class JMessageImpl implements JMessage {
  constructor(private message: Message) { }

  get messageKey(): MessageKey {
    return {
      channelId: this.message.channelId,
      messageId: this.message.id,
    }
  }

  get channelId() {
    return this.message.channelId
  }

  get author() {
    return this.message.author
  }

  get content() {
    return this.message.content
  }

  get clearReactions(): Observable<JMessage> {
    return defer(() => of(this.message.reactions)).pipe(
      mergeMap(reactionManager => reactionManager.removeAll()),
      map(resultingMessage => new JMessageImpl(resultingMessage)),
    )
  }

  get reactions(): Observable<JReaction> {
    const reactions = defer(() => rxFrom(this.message.awaitReactions({
      filter: (reaction, user) => {
        const emojiName = reaction.emoji.name
        if (emojiName == null) {
          return false
        }

        if (user.bot) {
          return false
        }

        const reactions = [
          QueueReactions.TWO_BACK,
          QueueReactions.BACK,
          QueueReactions.NEXT,
          QueueReactions.TWO_NEXT,
        ]

        const reactionsAsString = reactions.map(r => `${r}`)
        return reactionsAsString.includes(emojiName)
      },
      max: 1,
      time: 60000,
      errors: ['time'],
    })))

    const firstReaction = reactions.pipe(
      filter(reactions => reactions.size > 0),
      mergeMap(reactions => [reactions.first()]),
      mergeMap(reaction => reaction !== undefined ? [reaction] : []),
    )

    return firstReaction.pipe(
      map(reaction => reaction.emoji.name),
      mergeMap(emojiReaction => emojiReaction !== null ? [emojiReaction] : []),
    )
  }

  edit(update: MessageContent): Observable<JMessage> {
    return defer(() => rxFrom(this.message.edit(update))).pipe(
      map(updatedMsg => new JMessageImpl(updatedMsg)),
    )
  }

  react(reaction: string): Observable<JMessage> {
    return defer(() => rxFrom(this.message.react(reaction))).pipe(
      mapTo(this),
    )
  }

  send(message: MessageContent): Observable<JMessage> {
    return defer(() => rxFrom(this.message.channel.send(message))).pipe(
      map(msg => new JMessageImpl(msg)),
    )
  }
}

export const from = (message: Message): JMessage => new JMessageImpl(message)
