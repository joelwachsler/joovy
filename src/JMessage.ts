import { Message } from 'discord.js';

export interface JMessage {
  channelId: string
  author: {
    bot: boolean
    id: string
  }
  content: string
}

export namespace JMessage {
  export const from = (message: Message): JMessage => message
}
