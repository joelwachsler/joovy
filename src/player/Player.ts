import { AudioPlayer, createAudioPlayer, createAudioResource, joinVoiceChannel, VoiceConnection, VoiceConnectionStatus } from '@discordjs/voice'
import { Message, VoiceChannel } from 'discord.js'
import { map, Observable } from 'rxjs'
import { logger } from '../logger'
import { Ytdl } from './Ytdl'

export interface Player {
  play(track: Player.Track): Observable<void>
  disconnect(): void
}

export namespace Player {
  export interface Track {
    name: string
    link: string
  }

  export type Factory = Observable<Player>

  export const from = (message: Message): Observable<Player> => {
    const throwError = (err: string) => {
      throw Error(`Unable to join voice channel: ${err}`)
    }

    const voiceChannel = message.member?.voice.channel ?? throwError('Could not determine voice channel')
    logger.info(`Joining channel: ${voiceChannel?.id}...`)

    const connection = joinVoiceChannel({
      channelId: voiceChannel.id,
      guildId: voiceChannel.guildId,
      adapterCreator: voiceChannel.guild.voiceAdapterCreator,
    })

    const player = createAudioPlayer()
    connection.subscribe(player)

    return new Observable(observer => {
      connection.once(VoiceConnectionStatus.Ready, () => {
        logger.info('Connection is ready')

        if (voiceChannel instanceof VoiceChannel) {
          observer.next(new PlayerImpl(player, connection))
          observer.complete()
        } else {
          observer.error(`Voice channel was not correct type, got: ${typeof voiceChannel}, expected: ${typeof VoiceChannel}`)
        }
      })
    })
  }

  class PlayerImpl implements Player {
    private bass = 3
    private volume = 0.25
    private begin = 0

    constructor(private player: AudioPlayer, private connection: VoiceConnection) { }

    play(track: Track) {
      return this.createReadStream(track)
        .pipe(
          map(dl => createAudioResource(dl)),
          map(resource => this.player.play(resource)),
        )
    }

    private createReadStream(track: Track) {
      return Ytdl.createStream({
        url: track.link,
        options: {
          encoderArgs: ['-af', `bass=g=${this.bass},dynaudnorm=f=200,volume=${this.volume}`],
          opusEncoded: true,
          seek: (this.begin ?? 0) / 1000,
        },
        ytdlOptions: {
          filter: 'audioonly',
          quality: 'highestaudio',
          highWaterMark: 1 << 25,
        },
      })
    }

    disconnect() {
      this.player.stop()
      this.connection.disconnect()
    }
  }
}
