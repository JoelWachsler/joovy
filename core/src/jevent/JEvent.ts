import { Client, Message, TextChannel } from 'discord.js'
import { defer, from, map, mergeMap, Observable, of } from 'rxjs'
import JMessage, { from as jMessageFrom, MessageKey } from '../JMessage'
import { EventStore } from './EventStore'
import { Factory } from './Factory'
import { WithBaseFunctionality } from './mixin/BaseFunctionality'
import WithFactory from './mixin/Factory'
import WithSendMessage from './mixin/SendMessage'
import { ResultFactory } from './Result'
import { SendMessage } from './SendMessage'

/**
 * "JEvent" or "Joovy Event" represents an interaction a user or bot has created.
 * 
 * The main reason for this interface is to make it easier to create "mock"
 * implementations of a user interaction.
 */
export default interface JEvent extends ResultFactory, Factory, EventStore, SendMessage {
  readonly message: JMessage
  readonly timestamp: number,
}

const messageStore = new Map<string, Map<string, unknown>>()

const getOrCreateStore = (id: string) => {
  let store = messageStore.get(id)
  if (!store) {
    store = new Map()
    messageStore.set(id, store)
  }

  return store
}

export const removeObjectStore = (event: JEvent) => {
  return defer(() => {
    const channelId = event.message.channelId
    messageStore.delete(channelId)

    return event.result({ store: { event: 'removed', id: channelId } })
  })
}

export const fromMessage = (message: Observable<Message>): Observable<JEvent> => {
  return message.pipe(
    map(message => WithFactory(WithBaseFunctionality(jMessageFrom(message), () => getOrCreateStore(message.channelId), message.createdTimestamp), message)),
    map(EventClass => WithSendMessage(EventClass)),
    map(EventClass => new EventClass()),
  )
}

export const fromMessageKey = (client: Client, messageKey: MessageKey) => {
  return defer(() => from(client.channels.fetch(messageKey.channelId))).pipe(
    mergeMap(channel => channel ? [channel] : []),
    mergeMap(channel => channel instanceof TextChannel ? [channel] : []),
    mergeMap(textChannel => textChannel.messages.fetch()),
    map(messages => messages.get(messageKey.messageId)),
    mergeMap(message => message ? [message] : []),
    mergeMap(message => fromMessage(of(message))),
  )
}

// eslint-disable-next-line @typescript-eslint/ban-types
export type Constructor<T = {}> = new (...args: any[]) => T
export type BaseConstructor<T extends JMessage = JMessage> = Constructor<{ message: T }>

export const Base = <T extends JMessage = JMessage>(message: T, timestamp: number) => {
  return class {
    message = message
    timestamp = timestamp
  }
}