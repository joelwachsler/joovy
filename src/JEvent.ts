import { Message } from 'discord.js'
import { map, Observable } from 'rxjs'
import { JMessage } from './JMessage'
import * as Player from './player/Player'
import { getOrCreateObjectStore, getOrCreateStringStore, ObjectStore, StringStore } from './Store'

export default interface JEvent {
  readonly store: EventStore
  readonly factory: Factory
  readonly message: JMessage
  withResult(resultToAdd: any): this
  readonly result: any[]
}

export const from = (message$: Observable<Message>): Observable<JEvent> => {
  return message$.pipe(
    map(message => Base(message)),
    map(EventClass => WithStore(EventClass)),
    map(EventClass => WithResult(EventClass)),
    map(EventClass => WithFactory(EventClass)),
    map(EventClass => new EventClass()),
  )
}

interface EventStore {
  string: Observable<StringStore>
  object: Observable<ObjectStore>
}

interface Factory {
  player: Player.Factory
}

// eslint-disable-next-line @typescript-eslint/ban-types
type Constructor<T = {}> = new (...args: any[]) => T
type BaseConstructor<T extends JMessage = Message> = Constructor<{ message: T }>

const Base = (message: Message) => {
  return class {
    message = message
  }
}

export const WithStore = <TBase extends BaseConstructor<JMessage>>(Base: TBase) => {
  return class extends Base implements EventStore {
    get string() {
      return getOrCreateStringStore(this.message)
    }

    get object() {
      return getOrCreateObjectStore(this.message)
    }

    get store() {
      return this
    }
  }
}

export const WithFactory = <TBase extends BaseConstructor>(Base: TBase) => {
  return class extends Base implements Factory {
    get player() {
      return Player.from(this.message)
    }

    get factory() {
      return this
    }
  }
}

export const WithResult = <TBase extends Constructor>(Base: TBase) => {
  return class extends Base {
    private _result: any[] = []

    get result() {
      return this._result
    }

    /**
     * Needed to prevent jest matcher errors.
     */
    set result(result: any[]) {
      this._result = result
    }

    withResult(resultToAdd: any) {
      this.result.push(resultToAdd)
      return this
    }
  }
}
