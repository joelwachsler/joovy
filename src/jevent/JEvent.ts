import { Message, MessageEmbed } from 'discord.js'
import { delay, map, Observable } from 'rxjs'
import { JMessage } from '../JMessage'
import * as Player from '../player/Player'
import { ObjectStore, StoreProvider, StringStore } from '../Store'
import WithEventStore from './impl/EventStore'
import WithFactory from './impl/Factory'
import WithResult from './impl/Result'
import WithSendMessage from './impl/SendMessage'

export default interface JEvent extends Result, Factory, EventStore, SendMessage {
  readonly message: JMessage
}

export const from = (message$: Observable<Message>): Observable<JEvent> => {
  const store = new Map<string, unknown>()

  return message$.pipe(
    map(message => WithBaseFunctionality(message, () => store)),
    map(EventClass => WithFactory(EventClass)),
    map(EventClass => WithSendMessage(EventClass)),
    map(EventClass => new EventClass()),
  )
}

// eslint-disable-next-line @typescript-eslint/ban-types
export type Constructor<T = {}> = new (...args: any[]) => T
export type BaseConstructor<T extends JMessage = JMessage> = Constructor<{ message: T }>

const Base = <T extends JMessage = JMessage>(message: T) => {
  return class {
    message = message
  }
}

export const WithBaseFunctionality = <T extends JMessage = JMessage>(message: T, storeProvider: StoreProvider) => {
  return WithResult(WithEventStore(Base(message), storeProvider))
}

export interface EventStore {
  readonly store: {
    readonly string: Observable<StringStore>
    readonly object: Observable<ObjectStore>
  }
}

export const delayFactoryImpl = <T>(ms: number) => delay<T>(ms)

export type DelayFactory = typeof delayFactoryImpl

export interface Factory {
  readonly factory: {
    readonly player: Player.Factory
    readonly delay: DelayFactory
  }
}

export type ResultResult = string | Record<string, unknown>

export type ResultArg<T = undefined> = { result: ResultResult, item: T }

export interface Result {
  result(resultToAdd: ResultResult, ...andThen: Observable<ResultEntry>[]): Observable<ResultEntry>
  complexResult<T = undefined>(arg: ResultArg<T>, ...andThen: Observable<ResultEntry>[]): Observable<ResultEntry<T>>
}

export interface ResultEntry<T = any> {
  item: T,
  result: ResultResult
  event: JEvent
}

export interface SendMessage {
  sendMessage(message: string | MessageEmbed): Observable<ResultEntry>
}
