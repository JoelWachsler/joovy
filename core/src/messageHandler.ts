import { Client } from 'discord.js'
import { catchError, filter, mergeMap, Observable } from 'rxjs'
import { handle } from './commands/command'
import { errorHandler } from './errorHandler'
import JEvent, { fromMessageKey } from './jevent/JEvent'
import { EmptyResult, Result } from './jevent/Result'
import { createConsumer, createProducer, KMessage, Topics } from './kafka/kafka'
import { logResult } from './resultLogger'

export const sendMessageEvent = (event: Observable<JEvent>): Observable<Result> => {
  return createProducer().pipe(mergeMap(producer => {
    return event.pipe(mergeMap(event => {
      return producer.send({
        event,
        message: createMessageEvent(event),
        topic: Topics.NewMessage,
      })
    }))
  }))
}

export const handleMessageEvents = (client: Client) => {
  const messageHandling = createConsumer<MessageEvent>({ groupId: 'joovy-message-handler', topic: Topics.NewMessage }).pipe(
    mergeMap(msg => fromMessageKey(client, msg.meta.messageKey)),
    handleMessage,
  )

  logResult('handleMessageEvents', messageHandling)
}

export const handleMessage = (event: Observable<JEvent>): Observable<Result> => {
  return event.pipe(
    mergeMap(event => {
      const message = event.message
      if (message.author.bot) {
        return event.result({ ignored: `${message.content} was sent by a bot` })
      } else if (!message.content.startsWith('/')) {
        return event.result({ ignored: `${message.content} does not start with a slash` })
      } else {
        return handle(event).pipe(
          filter(r => !(r instanceof EmptyResult)),
          catchError(err => errorHandler(event, err)),
        )
      }
    }),
  )
}

const createMessageEvent = (event: JEvent): MessageEvent => {
  return {
    meta: {
      messageKey: event.message.messageKey,
      name: 'message-event',
      version: '1',
    },
    content: event.message.content,
  }
}

type MessageEvent = KMessage<'message-event', '1'> & { content: string }
