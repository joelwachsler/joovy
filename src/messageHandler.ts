import { mergeMap, Observable, of } from 'rxjs'
import { handle } from './commands/command'
import JEvent from './JEvent'

export const handleMessage = (event$: Observable<JEvent>): Observable<JEvent> => {
  return event$.pipe(
    mergeMap(event => {
      const message = event.message
      if (message.author.bot) {
        return of(event.withResult({ ignored: `${message.content} was sent by a bot` }))
      } else if (!message.content.startsWith('/')) {
        return of(event.withResult({ ignored: `${message.content} does not start with a slash` }))
      } else {
        return handle(event)
      }
    }),
  )
}
