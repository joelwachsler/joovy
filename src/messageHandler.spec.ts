import { Observable, of } from 'rxjs'
import { TestScheduler } from 'rxjs/testing'
import JEvent, { WithResult, WithStore } from './JEvent'
import { JMessage } from './JMessage'
import { handleMessage } from './messageHandler'
import Player, { Track } from './player/Player'

export const createScheduler = () => {
  return new TestScheduler((actual, expected) => expect(actual).toMatchObject(expected))
}

const scheduler = createScheduler()

const createTestEvent = (input?: Partial<JMessage>): JEvent => {
  class PlayerFake implements Player {
    play(_: Track): Observable<void> {
      throw new Error('Method not implemented.')
    }

    disconnect(): void {
      throw new Error('Method not implemented.')
    }
  }

  class EventFakeBase {
    get message() {
      return {
        author: {
          bot: false,
          id: 'testAuthorId',
        },
        channelId: 'testChannelId',
        content: 'testContent',
        ...input,
      }
    }

    get factory() {
      return {
        player: of(new PlayerFake()),
      }
    }
  }

  return new class EventFake extends WithStore(WithResult(EventFakeBase)) { }
}

test('should ignore bot messages', () =>  {
  scheduler.run(({ expectObservable, hot }) => {
    const event = createTestEvent({
      author: {
        bot: true,
        id: 'testAuthorId',
      },
      content: '/test',
    })

    const source$ = hot<JEvent>('a', { a: event })
    expectObservable(handleMessage(source$)).toBe('r', {
      r: {
        result: [
          { ignored: '/test was sent by a bot' },
        ],
      },
    })
  })
})

test('should ignore messages not starting with a slash', () => {
  scheduler.run(({ expectObservable, hot }) => {
    const event = createTestEvent({
      content: 'test',
    })

    const source$ = hot<JEvent>('a', { a: event })
    expectObservable(handleMessage(source$)).toBe('r', {
      r: {
        result: [
          { ignored: 'test does not start with a slash' },
        ],
      },
    })
  })
})

test('should join channel if not previously joined', () => {
  scheduler.run(({ expectObservable, hot }) => {
    const event = createTestEvent({
      content: '/play test',
    })

    const source$ = hot<JEvent>('a', { a: event })
    expectObservable(handleMessage(source$)).toBe('r', {
      r: {
        result: [
          { player: { joined: 'testing' } },
        ],
      },
    })
  })
})
