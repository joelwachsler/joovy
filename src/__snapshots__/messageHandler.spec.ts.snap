// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`disconnection disconnect should disconnect if connected to channel 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{\\"name\\":\\"/play test\\",\\"link\\":\\"test\\"} has been added to the queue",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "Now playing: {\\"name\\":\\"/play test\\",\\"link\\":\\"test\\"}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/disconnect",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "disconnected",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "C",
    },
  },
]
`;

exports[`player creation should create player if not previously created 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{\\"name\\":\\"/play test\\",\\"link\\":\\"test\\"} has been added to the queue",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "Now playing: {\\"name\\":\\"/play test\\",\\"link\\":\\"test\\"}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
]
`;

exports[`player creation should not create player if previously created 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{\\"name\\":\\"/play test\\",\\"link\\":\\"test\\"} has been added to the queue",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "Now playing: {\\"name\\":\\"/play test\\",\\"link\\":\\"test\\"}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{\\"name\\":\\"/play test2\\",\\"link\\":\\"test2\\"} has been added to the queue",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "Now playing: {\\"name\\":\\"/play test2\\",\\"link\\":\\"test2\\"}",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
]
`;

exports[`playlist should queue song if another one is already playing 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{\\"name\\":\\"/play test\\",\\"link\\":\\"test\\"} has been added to the queue",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "Now playing: {\\"name\\":\\"/play test\\",\\"link\\":\\"test\\"}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{\\"name\\":\\"/play test2\\",\\"link\\":\\"test2\\"} has been added to the queue",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{\\"name\\":\\"/play test3\\",\\"link\\":\\"test3\\"} has been added to the queue",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "Now playing: {\\"name\\":\\"/play test2\\",\\"link\\":\\"test2\\"}",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "Now playing: {\\"name\\":\\"/play test3\\",\\"link\\":\\"test3\\"}",
      },
    },
  },
  TestMessageValue {
    "frame": 15,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
]
`;
